### PR TITLE
17.0 tutorials abzo

### PR DESCRIPTION
**Pull request by Abdessamad-z**

### CODE REVIEW

File: estate/models/estate_property.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    best_price = fields.Float(compute="_compute_best_offer")
+
+    _sql_constraints = [
+        ('check_pos', 'CHECK(expected_price >= 0 AND selling_price >= 0)',
+         'Value must be positive.')
+    ]
+
+    @api.depends("garden_area", "living_area")
+    def _compute_total(self):
+        self.total_area = self.garden_area + self.living_area
```
Comment (vava-odoo) [inline estate/models/estate_property.py:60]: What if there are several records in `self`?

File: estate/models/estate_property.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    @api.depends("offer_ids.price")
+    def _compute_best_offer(self):
+        self.best_price = max([i.price for i in self.offer_ids] + [0])
```
Comment (vava-odoo) [inline estate/models/estate_property.py:64]: ORM has many tools to deal with recordset. One of them is `mapped`
```suggestion
        self.best_price = max(self.offer_ids.mapped('price'), default=0)
```
Also, this works only for one record in self. So either you start with `self.ensure_one()`, either you loop over several records

File: estate/models/estate_property.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    @api.onchange("garden")
+    def _onchange_partner_id(self):
+        if self.garden:
+            self.garden_area = 10
+            self.garden_orientation = 'north'
+        else:
+            self.garden_area = 0
+            self.garden_orientation = None
```
Comment (vava-odoo) [inline estate/models/estate_property.py:73]: We tend to prefer `False`, instead of `None`

File: estate/models/estate_property.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    def cancel(self):
+        for record in self:
+            if record.state == 'Sold':
+                raise UserError("Canceled property can not be sold")
+            record.state = "Canceled"
```
Comment (vava-odoo) [inline estate/models/estate_property.py:79]: ORM tip: prefer to batch operations:
```suggestion
        if 'Sold' in self.mapped('state'):
            raise UserError("Canceled property can not be sold")
        self.state = "Canceled"
```

File: estate/models/estate_property.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    @api.onchange("garden")
+    def _onchange_partner_id(self):
+        if self.garden:
+            self.garden_area = 10
+            self.garden_orientation = 'north'
+        else:
+            self.garden_area = 0
+            self.garden_orientation = None
+
+    def cancel(self):
```
Comment (vava-odoo) [inline estate/models/estate_property.py:75]: Naming convention: start the name of an action with `action_`

File: estate/models/offer.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
+from odoo.exceptions import UserError
+from odoo import fields, models, api
+from odoo.tools import add
```
Comment (vava-odoo) [inline estate/models/offer.py:3]: https://www.odoo.com/documentation/master/contributing/development/coding_guidelines.html#imports

File: estate/models/offer.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+class Offer(models.Model):
+    _name = "offer"
+    _description = "Offer"
+    price = fields.Float(string="Price")
+    status = fields.Selection(copy=False, string="Status", selection=[("accepted", "Accepted"), ("refused", "Refused")],
+                              required=True)
+    partner_id = fields.Many2one("res.partner", string="Partner", required=True)
+    property_id = fields.Many2one("estate_property", string="Estate Property", required=True)
+    validity = fields.Integer(string="Validity", default="7")
```
Comment (vava-odoo) [inline estate/models/offer.py:14]: `string` could be clearer: what is the Unit of Measure? Month?
Also, the default should be of the good type, so 7, not "7"

File: estate/models/offer.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    @api.depends("validity")
+    def _compute_deadline(self):
+        for record in self:
+            record.date_deadline = add(record.create_date or fields.Date.today(),
+                                       days=record.validity)
+
+    def _inverse_total(self):
+        for record in self:
+            record.validity = (record.date_deadline - record.create_date.date()).days
```
Comment (vava-odoo) [inline estate/models/offer.py:30]: What if create_date is not provided?

File: estate/models/offer.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    def accept(self):
+        for record in self:
+            for i in record.property_id.offer_ids:
+                if i.status == "accepted":
```
Comment (vava-odoo) [inline estate/models/offer.py:35]: batch

File: estate/models/offer.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                    raise UserError("An offer was already accepted")
+            record.status = "accepted"
+            record.property_id.selling_price = self.price
+            record.property_id.buyer_id = self.partner_id.id
```
Comment (vava-odoo) [inline estate/models/offer.py:39]: what about other offers for the same property? You'd want to refuse them

File: estate/models/offer.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+        return True
+
+    def refuse(self):
+        for record in self:
+            record.status = "refused"
+            record.property_id.selling_price = 0
+            record.property_id.buyer_id = None
```
Comment (vava-odoo) [inline estate/models/offer.py:47]: Why? What if another offer is already accepted?

File: estate/views/estate_menus.xml [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+    <menuitem id="test_menu_root" name="Test">
+        <menuitem id="test_first_level_menu" name="List">
+            <menuitem id="test_model_menu_action" action="estate_property_action"/>
+        </menuitem>
+        <menuitem id="setting_level_menu" name="Setting">
+            <menuitem id="setting_type_menu_action" action="estate_property_type_action"/>
+        </menuitem>
+        <menuitem id="setting_level_menu" name="Setting">
+            <menuitem id="setting_tag_menu_action" action="estate_property_tag_action"/>
+        </menuitem>
```
Comment (vava-odoo) [inline estate/views/estate_menus.xml:12]: You can't have two times the same id. Instead
```suggestion
        <menuitem id="setting_level_menu" name="Setting">
            <menuitem id="setting_type_menu_action" action="estate_property_type_action"/>
            <menuitem id="setting_tag_menu_action" action="estate_property_tag_action"/>
        </menuitem>
```

File: estate/views/estate_menus.xml [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+    </menuitem>
+</odoo>
```
Comment (vava-odoo) [inline estate/views/estate_menus.xml:15]: Missing end-of-file empty line

File: estate/views/estate_proprety_views.xml [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+                                <group>
+                                    <field name="description"/>
+                                    <field name="bedrooms"/>
+                                    <field name="living_area"/>
+                                    <field name="facades"/>
+                                    <field name="garage"/>
+                                    <field name="garden"/>
+                                    <field name="garden_area"/>
+                                    <field name="garden_orientation"/>
+                                    <field name="property_type_id" string="Property Type"/>
```
Comment (vava-odoo) [inline estate/views/estate_proprety_views.xml:66]: By default, the string is the same as the one defined on the field, so no need for this one

**Review by vava-odoo — COMMENTED** [review]
Good! Continue like this, almost the end of it :muscle: 
Could you remove all the non-necessary files? :eyes:

**Comment by Abdessamad-z** [comment]
Done thanks for the review.

**Pull request closed without merging**